### ISSUE TITLE
Memory efficient BLS with input-dependent number of inference requests

### ISSUE THREAD

**Issue by razorx89** [triton-inference-server/server#3807]
**Description**
I am trying to implement a sliding window inference approach via BLS in the python backend for 3D medical image segmentation. The whole point of this is to reduce network traffic, especially with larger models with many output classes the transfer of probability maps back to the client for aggregation between overlapping crops can be quite expensive. This script currently works, but steadily increases the allocated shared memory. I also noticed this statement in the README:

> The number of inference requests that can be executed as a part of your model execution is limited to the amount of shared memory available to the Triton server. If you are using Docker to start the TritonServer, you can control the shared memory usage using the --shm-size flag.

I wonder why all the inference requests and responses must be kept in memory during an ongoing python model request? In my case I am consuming the result of the response and calculate a totally different output for my model. Aggregation happens with probability maps, but the result is a `uint8` array with predicted class indices. For a CT scan with 350 slices, crop size of 32 slices and overlap of 75%, a shared memory of 8GB is fully consumed when the model is just half way through. Sure, increasing the shared memory solves the problem, but the memory footprint is totally input dependent (number of slices) and in theory should be constant (for the overlapping inference algorithm, the output array is of course also dependent on the input). I also tried to explicitly `del infer_response` and `del infer_request` with a `gc.collect()` but the memory keeps increasing.

**Triton Information**
nvcr.io/nvidia/tritonserver:21.12-py3

**To Reproduce**

```python
class TritonPythonModel:
    def initialize(self, args):
        # ...
  
    def handle_request(self, request):
        arr = pb_utils.get_input_tensor_by_name(request, "input").as_numpy()
        result = np.empty(shape=arr.shape, dtype=np.uint8)
    
        # ... compute crop positions
    
        # send crops to model
        for crop_idx in crop_indices:
            slc = tuple(slice(None) if i != self.axis else slice(crop_idx, crop_idx + self.mdim_size) for i in range(arr.ndim))   
            crop = arr[slc]
            infer_request = pb_utils.InferenceRequest(
                model_name=self.requested_model_name,
                requested_output_names=[self.requested_model_output_name],
                inputs=[pb_utils.Tensor(self.requested_model_input_name, crop[np.newaxis, ..., np.newaxis])]
            )
            infer_response = infer_request.exec()
    
            # ...
            
            crop_pred = pb_utils.get_output_tensor_by_name(infer_response, self.requested_model_output_name).as_numpy()
    
            # ... keep intermediate results in buffer and 
            # ... assign final results to the result array
    
        return pb_utils.InferenceResponse(
            output_tensors=[
                pb_utils.Tensor("output", result)
            ],
        )
```

**Expected behavior**
Shared memory should not increase in this case, because the result of inference requests is fully consumed, references to the inference request and response objects were dropped, and not needed for the final python model response.

**Comment by Tabrizian** [comment]
In the current version of python backend the shared memory allocated during the request execution will not be deallocated until the execution of the batch of requests is finished. Improving the shared memory usage is on our roadmap.

**Comment by Tabrizian** [comment]
This has been fixed in the 22.04 release.
